src/slice.c

```diff
@@ -237,7 +295,7 @@ static SEXP vec_slice_impl(SEXP x, SEXP index) {
       Rf_setAttrib(out, R_NamesSymbol, names);
     }
 
-    out = vec_restore(out, x, index);
+    out = vec_restore(out, x, restore_size);
```

Comment:
This also fixes a small bug that I think was a holdover from when `vec_restore()` took `i` not `n`.